### PR TITLE
AP_MotorsMatrix: de-obfuscate final throttle calculation

### PR DESCRIPTION
**Pull request by 117-MasterChief**
The final throttle calculation in `AP_MotorsMatrix::output_armed_stabilizing` is written in a roundabout way that is hard to parse. This small rewrite does not materially change any calculation, but simply makes the final throttle calculation much easier to read and understand.

Since a few unnecessary intermediate variables are removed and utilization of IIFE, this is potentially slightly more efficient (I doubt this though), but a run through godbolt (or similar) with the right compiler settings is needed to determine that.

### CODE REVIEW

File: libraries/AP_Motors/AP_MotorsMatrix.cpp [hunk 1]
```diff
@@ -368,41 +368,43 @@ void AP_MotorsMatrix::output_armed_stabilizing()
...
     rpy_low *= rpy_scale;
-    throttle_thrust_best_rpy = -rpy_low;
-    float thr_adj = throttle_thrust - throttle_thrust_best_rpy;
-    if (rpy_scale < 1.0f) {
-        // Full range is being used by roll, pitch, and yaw.
-        limit.roll = true;
-        limit.pitch = true;
-        limit.yaw = true;
-        if (thr_adj > 0.0f) {
+    const float final_throttle = [&]() {
```
Comment (khancyr) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp:371]: hahahaaha I am really curious about what the other will say about this.
 In this case, I don't think this will bring much improvements as the compiler should already handle this simple variable.

**Review by khancyr — CHANGES_REQUESTED** [review]
Change is mostly reordering, no changes in the calculus.

I let the maintainers deal with accepting the IIFE or not. That would be the first non STIL occurance I think, but that isn't a bad thing IMHO

@117-MasterChief  can you rebase to squash your commits and get ride of the merge one's ? Thanks

File: libraries/AP_Motors/AP_MotorsMatrix.cpp (hunk 1 above)
Comment (117-MasterChief) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp:371]: yes, i dont expect any improvements. possibly no changes at all. but the compiler works in mysterious ways...

**Comment by 117-MasterChief** [comment]
@khancyr

> I let the maintainers deal with accepting the IIFE or not. That would be the first non STIL occurance I think, but that isn't a bad thing IMHO

This should compile similarly to current impl, if not exactly.

> can you rebase to squash your commits and get ride of the merge one's ? Thanks

Sure, but can I ask why? PRs when merged into master should be automatically squashed into 1 commit. Is that not the case here? Or is there another reason?

**Review by lthall — COMMENTED** [review]
Nope. I will not support this change.
I appreciate your feedback and I like some aspects of your suggestion. I have some other changes in this part of the code to improve clarity and readability and will consider what you have suggested here.

**Comment by 117-MasterChief** [comment]
@lthall 

> Nope.

Are you responding to me or @khancyr , or is this a way of rejecting this PR?

**Comment by 117-MasterChief** [comment]
@magicrub for viz, since you reviewed my other PR modifying the same file.

**Comment by lthall** [comment]
>Are you responding to me or @khancyr , or is this a way of rejecting this PR?

I added some clarification to my reply above.

I know my feedback on your PR is not going to be received in a productive way and end up just waisting my time and energy. So yes, this is a rejection of your PR.

**Comment by IamPete1** [comment]
Not a fan of the lambda function I think it makes the code harder to follow, especially for those less well versed c++. If you can make the maths easier to follow without that then it might be worth it, I agree the `thr_adj` stuff is confusing. If we can find a better structure then we need to prove the result is the same, for this we have the motors example, you can run a sweep with `./libraries/AP_Motors/examples/AP_Motors_test/MotorTestSweep.sh` (it might be broken currently you have to go a and disable ESC telem I think). That will take ages to run and produce a bunch of files. Commit them to the tree and then add a commit reverting your changes run again and there should be no difference. 

The tests take ages to run, might be better get agreement on the cleaner function before bothering with the tests. 

Surprised we didn't also get a extra comment explaining the magic 0.5. ;)

**Comment by 117-MasterChief** [comment]
> I know my feedback on your PR is not going to be received in a productive way and end up just waisting my time and energy. So yes, this is a rejection of your PR.

well, i am sorry you took our past disagreements personally and are maintaining grudges. in my last PR i did incorporate you and @IamPete1 feedback in the closing comments of the PR.

> I have some other changes in this part of the code to improve clarity and readability and will consider what you have suggested here.

OK, but I hope though that you dont simply subsume these changes into your own. I think that would be unethical. but if you have a branch containing your upcoming changes, I would like to see it.

@khancyr i guess this might answer your question :)

**Comment by 117-MasterChief** [comment]
@IamPete1 in general, IIFE is a great method for complex initialization and occasionally has performance benefits. it has minimal syntactical overhead and should never compile worse than what its replacing. its already used in SITL code, according to @khancyr . Other devs should get become familiar...

**Comment by 117-MasterChief** [comment]
Closed because Leonard no likey

**Comment by khancyr** [comment]
nothing personal here. The most important change here is IIFE and that something that don't bring benefit here but bring some complexity for those not used to lambda. 
You will see many place in the code where we aren't using true CPP semantics as it could be complex for some : "smart_pointers", template , etc.

**Comment by 117-MasterChief** [comment]
@khancyr thanks for genuine feedback. I believe you are correct, except look at @lthall 's comments. It's personal and emotional (angry).

> For the PR style...

Got it.

**Comment by 117-MasterChief** [comment]
Closed because Leonard is not familiar with IIFE.
